src/common/memwritestream.cpp

```diff
@@ -114,12 +128,14 @@ void MemoryWriteStreamDynamic::ensureCapacity(size_t newLen) {
 size_t MemoryWriteStreamDynamic::write(const void *dataPtr, size_t dataSize) {
 	assert(dataPtr);
 
-	ensureCapacity(_size + dataSize);
+	ensureCapacity(pos() + dataSize);
 
 	std::memcpy(_ptr, dataPtr, dataSize);
 
+	if (pos() + dataSize > _size)
+		_size += dataSize;
```

Comment:
Hmm, are you sure that's correct? Unless I'm not seeing something, shouldn't that be `_size += dataSize - pos();`? I.e. `_size` should grow by the difference/overhang.

If you could also add a unit test for that case then, that would be great. Something like write 6 bytes, make sure that the size of the stream is 6, seek back two bytes, write another 6 bytes and make sure that the size of the stream is then 10 (instead of 12), and that the contents is also still correct.

And the same test for the file writer, too.

Likewise, please add a test to the non-dynamic write stream that wants to write more. Something like I said above, write 6 bytes, seek back 2, write another 6 and make sure that the method returns 2 (for 2 bytes written, because there's only place in the buffer for 2). And content matching.